src/core_plugins/console/public/src/autocomplete/components/url_pattern_matcher.js

```diff
@@ -0,0 +1,133 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import _ from 'lodash';
+import {
+  SharedComponent,
+  ConstantComponent,
+  AcceptEndpointComponent,
+  ListComponent,
+  SimpleParamComponent,
+} from '.';
+
+/**
+ * @param parametrizedComponentFactories a dict of the following structure
+ * that will be used as a fall back for pattern parameters (i.e.: {indices})
+ * {
+ *   indices: function (part, parent) {
+ *      return new SharedComponent(part, parent)
+ *   }
+ * }
+ * @constructor
+ */
+export class UrlPatternMatcher {
+  // This is not really a component, just a handy container to make iteration logic simpler
+  constructor(parametrizedComponentFactories) {
+    ['HEAD', 'GET', 'PUT', 'POST', 'DELETE'].forEach((method) => {
```

Comment:
It might be nice to leave a breadcrumb trail to help others who come through this code. I'd benefit from a hint on this line:

```js
// We'll group endpoints by the methods which are attached to them, to avoid suggesting
// endpoints that are incompatible with the method that the user has entered.
```